inst/pages/exercises.qmd

```diff
@@ -1674,71 +1692,134 @@ p <- Heatmap(
+
+```{r}
+#| label: extra-kmeansclustering-heatmap-ex2
+#| eval: FALSE
+#| code-fold: true
+#| code-summary: "Show solution"
 
-```{r, eval = FALSE}
+library(mia)
+library(SummarizedExperiment)
+library(vegan)
 library(scater)
 
-# Setting seed for reproducibility
-set.seed(123)
+# Load the dataset
+data(GlobalPatterns)
+tse <- GlobalPatterns
 
-# Perform k-means clustering on the reduced dimensions of pcoa_bray
-pcoa_bray_clusters <- kmeans(reducedDim(pcoa_bray), centers = 3)$cluster
+# Perform MDS on the abundance data using scater::runMDS
+tse <- runMDS(tse, FUN = vegan::vegdist, method = "bray", assay.type = "counts")
 
-# Plot the reduced dimensions of pcoa_bray with clusters as colors
-plotReducedDim(pcoa_bray, colour_by = pcoa_bray_clusters)
+# Extract the reduced dimensions
+mds_data <- reducedDim(tse, "MDS")
 
-# Perform k-means clustering on the reduced dimensions of pca_aitchison
-pca_aitchison_clusters <- kmeans(reducedDim(pca_aitchison), centers = 3)$cluster
+# Perform k-means clustering (k = 3) using the base R kmeans function
+set.seed(123)
+kmeans_result <- kmeans(mds_data, centers = 3)
 
-# Plot the reduced dimensions of pca_aitchison with clusters as colors
-plotReducedDim(pca_aitchison, colour_by = pca_aitchison_clusters)
-```
+# Add cluster labels to the SummarizedExperiment object
+# Convert clustering result to a factor
+tse$cluster <- factor(kmeans_result$cluster)
 
-Where: 
-- `reducedDim` is a function that extracts the reduced dimensionality
-representation of your data (e.g., from a PCA or PCoA object).
-- `plotReducedDim` is a function that can plot these reduced dimensions and
-has a 'colour_by argument' that accepts a vector of cluster assignments.
+# Visualize MDS with clusters
+plotReducedDim(tse, dimred = "MDS", colour_by = "cluster") +
+  labs(title = "MDS with k-means Clustering") +
+  theme_minimal()
+```
 
-#### 2. Heatmap with Annotations:
+3. Hierarchial Clustering
+- Hierarchical clustering of samples based on abundance data.
+- Visualizes the clustered data using a heatmap, with cluster information added to metadata.
 
-- Annotate the heatmap with additional metadata (e.g., subject groups or
-sample types).
+```{r}
+#| label: extra-hierarchialclustering-heatmap-ex3
+#| eval: FALSE
+#| code-fold: true
+#| code-summary: "Show solution"
 
-```{r, eval = FALSE}
-# Load necessary libraries
-library(SummarizedExperiment)
 library(ComplexHeatmap)
 
-# Example data creation
-# Assuming enterotype is a SummarizedExperiment object with sample metadata
-# and assay_data is a matrix containing assay data
+# Perform hierarchical clustering on the MDS data (extracted in last exercise)
+dist_matrix <- dist(mds_data)  # Compute distance matrix on MDS dimensions
+hclust_result <- hclust(dist_matrix, method = "complete")
+
+# Cut the dendrogram into 3 clusters
+cluster_labels <- cutree(hclust_result, k = 3)
+
+# Add cluster labels to the SummarizedExperiment object
+colData(tse)$cluster <- factor(cluster_labels)
+
```

Comment:
We have mia::addCluster that can perform these operations with a single line. Could you switch to that?

Reply:
addCluster() is generally used with predefined clustering methods in the bluster package, such as KmeansParam(I switched the previous exercise to using addCluster) but there is no predefined parameter for hierarchical clustering, so addCluster does not directly fit this usage pattern. It is rather complicating it since I am having to define a new dummy BLUSPARAM, do you think it will be too complex that way?

Reply:
Ah, this is interesting. 

Quick search shows that there are ready-made solutions for hierarchical clustering of SE objects:
1) [scran::quickCluster()](https://rdrr.io/bioc/scran/man/quickCluster.html)
2) [tomoda::hierarchClust](https://s3.jcloud.sjtu.edu.cn/899a892efef34b1b944a19981040f55b-oss01/bioconductor/3.14/bioc/manuals/tomoda/man/tomoda.pdf)
3) [clusterExperiment](https://www.bioconductor.org/packages/release/bioc/vignettes/clusterExperiment/inst/doc/clusterExperimentTutorial.html#plotHeatmap)

Perhaps some other solutions exists as well.

Could you check if some of these could provide a handy solution?

Reply:
Yes sure. It will be interesting to try new ways :))